File: backend/src/main/java/fr/pedalons/dto/notifications/request/NotificationPreferencesRequest.java
package fr.pedalons.dto.notifications.request;

import fr.pedalons.dto.validation.ValidateSchema;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import java.util.List;
import org.eclipse.microprofile.openapi.annotations.media.Schema;
import org.jspecify.annotations.Nullable;

/** A partial update: cells and teams not listed keep their value, as does an absent digest. */
@Schema(description = "Notification preference cells to change")
@ValidateSchema
public record NotificationPreferencesRequest(
    @Schema(description = "The cells to change", required = true) @NotNull @Size(max = 200)
        List<@Valid @NotNull NotificationPreferenceUpdate> preferences,
    @Schema(description = "The teams to mute or unmute") @Size(max = 200)
        @Nullable List<@Valid @NotNull NotificationTeamPreferenceUpdate> teams,
    @Schema(description = "Switch the daily e-mail digest on or off; absent leaves it")
        @Nullable Boolean emailDigest) {

  /** Cells only. */
  public NotificationPreferencesRequest(List<NotificationPreferenceUpdate> preferences) {
    this(preferences, null, null);
  }
}

File: frontend/src/api/dto/notificationPreferencesRequest.ts
import type { NotificationPreferenceUpdate } from './notificationPreferenceUpdate.ts'
import type { NotificationTeamPreferenceUpdate } from './notificationTeamPreferenceUpdate.ts'

/**
 * Notification preference cells to change
 */
export interface NotificationPreferencesRequest {
  /**
   * The cells to change
   * @maxItems 200
   */
  preferences: NotificationPreferenceUpdate[]
  /**
   * The teams to mute or unmute
   * @maxItems 200
   */
  teams?: NotificationTeamPreferenceUpdate[]
  /** Switch the daily e-mail digest on or off; absent leaves it */
  emailDigest?: boolean
}
